8337960: Improve performance of mfwrapper by reusing GStreamer media buffers for decoded video

sashamatveev opened the pull request:
- Added new class `CMFGSTBuffer` which can allocate memory internally or provide GStreamer allocated memory to Media Foundation.
- Added `GstBufferPool` to limit allocation of output buffers used for rendering (memory will not be allocated for each buffer, but instead will be reused from pool). Limits are 3 min buffers and 6 max buffers. During testing 3 buffers was enough.
- Changed `CoInitializeEx` to `COINIT_MULTITHREADED` as per Media Foundation requirements.
- Added error handling for `ProcessOutput` in case of https://bugs.openjdk.org/browse/JDK-8329227. With error handling `MediaPlayer` fails nicely instead of silent hang.

tiainen reviewed (approved):
All builds went fine.

kevinrushforth commented on modules/javafx.media/src/main/native/gstreamer/plugins/mfwrapper/mfgstbuffer.h:
```diff
@@ -0,0 +1,85 @@
+private:
+    HRESULT AllocateOrGetBuffer(BYTE **ppbBuffer);
+
+private:
```
Minor: isn't this redundant?

sashamatveev replied:
Yes, fixed.

kevinrushforth commented on modules/javafx.media/src/main/native/gstreamer/plugins/mfwrapper/mfgstbuffer.cpp:
```diff
@@ -0,0 +1,262 @@
+    if (ppbBuffer == NULL)
+        return E_INVALIDARG;
+
+    // If we have GStreamer get buffer cllback set, then call it to get
```
Minor typo: `cllback` --> `callback`

sashamatveev replied:
Fixed.

kevinrushforth commented on modules/javafx.media/src/main/native/gstreamer/plugins/mfwrapper/mfwrapper.cpp:
```diff
@@ -844,6 +886,118 @@ static HRESULT mfwrapper_init_colorconvert(GstMFWrapper *decoder,
+}
+
+// Gets max length of configured media buffer we using for final rendering from
+// decoder ot color convert.
```
Minor: "ot" --> "of" ?

sashamatveev replied:
Should be "or". Fixed.

kevinrushforth commented on modules/javafx.media/src/main/native/gstreamer/plugins/mfwrapper/mfgstbuffer.cpp:
```diff
@@ -0,0 +1,262 @@
+
+        (*ppbBuffer) = m_GstMapInfo.data;
+    }
+    // Lock can be called multiple times, so if we have GStreamer buffer
+    // allocated just return it.
```
Is it actually possible that Lock can be called multiple times? If it is, then I see that you don't keep track of the number of calls to Lock, so is it correct to assume that the caller does not match Lock / Unlock calls, but rather calls Unlock only once regardless of the number of calls to Lock?

sashamatveev replied:
Yes, based on documentation Lock() can be called multiple times to acquire pointer and it should be valid until last Unlock() is called. The caller is responsible to match Lock / Unlock calls. So, current implementation is not correct. I will fixed it. Also, based on documentation Lock() can be called from separate threads, so it should be thread safe. I will fix it as well.

kevinrushforth reviewed (approved):
The code looks reasonable. I left a few minor questions / comments. I'll reapprove if you make any changes.

All my testing looks good (I only tested on Windows, since this is in Windows-only code).

sashamatveev commented:
8337960: Improve performance of mfwrapper by reusing GStreamer media buffers for decoded video [v2]
- Fixed typos.
- Lock / Unlock is now thread safe and last Unlock will unmap buffer.
- AllocateOrGetBuffer() is updated to remap unmapped buffer and split into two logical sections.

kevinrushforth reviewed (approved):
This looks good to me now with the updated lock/unlock now handling multiple matched lock/unlock.

arapte reviewed (approved):
LGTM...

sashamatveev commented:
/integrate
